src-tauri/src/node/stream.rs

```diff
@@ -106,129 +252,120 @@ impl Serialize for StreamError {
-                                //     .blocking_send(StreamEvent::from_error(
-                                //         StreamError::IngestError(err),
-                                //         // @TODO: We should be able to get the operation causing the error.
-                                //         result.header,
-                                //     ))
-                                //     .expect("app_tx send");
-                                None
-                            }
-                        });
+impl<L, E> StreamControllerStore<L, E> for StreamMemoryStore<L, E>
+where
+    L: p2panda_store::LogId + Send + Sync,
+    E: p2panda_core::Extensions + Extension<L> + Send + Sync,
+{
+    type Error = StreamControllerError;
 
-                loop {
-                    match processor.next().await {
-                        Some(operation) => {
-                            // If the operation has a body we want to forward it to the application
-                            // layer as it might contain more information relevant for it.
-                            if let Some(body) = operation.body {
-                                app_tx
-                                    .send(StreamEvent::new(operation.header, body))
+    async fn ack(&self, operation_id: Hash) -> Result<(), Self::Error> {
+        let Ok(Some((header, _))) = self.operation_store.get_operation(operation_id).await else {
+            return Err(StreamControllerError::AckedUnknownOperation(operation_id));
+        };
+
+        let mut acked = self.acked.write().await;
+
+        let log_id: Option<L> = header.extension();
+        let Some(log_id) = log_id else {
+            return Err(StreamControllerError::MissingLogId(operation_id));
+        };
+
+        // Remember the "acknowledged" log-height for this log.
+        acked.insert((header.public_key, log_id), header.seq_num);
+
+        Ok(())
+    }
+
+    async fn unacked(
+        &self,
+        logs: HashMap<PublicKey, Vec<L>>,
+    ) -> Result<Vec<Operation<E>>, Self::Error> {
+        let acked = self.acked.read().await;
+
+        let mut result = Vec::new();
+        for (public_key, log_ids) in logs {
+            for log_id in log_ids {
+                match acked.get(&(public_key, log_id.clone())) {
+                    Some(ack_log_height) => {
+                        let Ok(operations) = self
+                            .operation_store
+                            .get_log(&public_key, &log_id, Some(*ack_log_height))
+                            .await;
+
+                        if let Some(operations) = operations {
+                            for (header, body) in operations {
+                                // @TODO(adz): Getting the encoded header bytes in a second
+                                // database call like this feels redundant and should be possible
+                                // to retreive just from calling "get_log".
```

Comment:
I'd create an issue for `p2panda_store` where we return the header bytes in all regular store methods